### ISSUE TITLE
helm upgrade with typo gets release stuck in status failed

### ISSUE THREAD

**Issue by thomas-riccardi** [helm/helm#3917]
# Scenario
- `helm install` chart, OK
- `helm upgrade --debug --dry-run`  OK
- `helm upgrade` chart with additional resources (e.g. a ServiceAccount) *and* a validation error in the definition of an already existing resource (e.g. type error in the container env value in a Deployment: `value: 1` instead of `value: "1"`)

## Result: 
- `helm upgrade` fails:
```
Error: UPGRADE FAILED: failed to create patch: failed to get versionedObject: unable to convert unstructured object to apps/v1beta1, Kind=Deployment: cannot convert int64 to string
```
- `helm status` fails:
```
LAST DEPLOYED: Fri Mar 30 12:16:16 2018
NAMESPACE: test
STATUS: FAILED
```
- subsequent `helm upgrade` fails too, with different error:
```
Error: UPGRADE FAILED: no ServiceAccount with the name "named-bumblebee-vulcan-api" found
```

## Workaround: 
- `helm rollback` to last successfully `DEPLOYED` revision, fix type error, re-upgrade

## Comment
There are multiple issues of strange failures on `helm upgrade` (#1193, #3785)/
This issue is a possible duplicate of #3785 but with a precise scenario (what tipped me off is [this comment](https://github.com/kubernetes/helm/issues/1193#issuecomment-358789168)).

It is hard to debug because 
- `helm status` gives no meaningful information (cf #3077).
- `helm upgrade --debug --dry-run` works: the resources definitions validation are not done in dry-run

# Info
## Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}
```
## Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.10-gke.0", GitCommit:"16ebd0de8e0ab2d1ef86d5b16ab1899b624a77cd", GitTreeState:"clean", BuildDate:"2018-03-20T20:21:01Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

## Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
GKE

**Comment by jascott1** [comment]
Hi @thomas-riccardi 

The chart has an error and then your next upgrade is trying to base of that error so its expected behavior of sorts although not too friendly. What are you expecting as a UX here?

**Comment by helgi** [comment]
@jascott1 I would expect to either base it off the last successful deploy *or* do a refresh (somehow, TF like) from Kubernetes and let that through if it ends up being identical.

**Comment by thomas-riccardi** [comment]
Hi @jascott1 ,
I'm not aware of the real constraints for upgrade (e.g. need to first roll-back to a successful state as suggested @helgi). However, as a helm user I just don't understand what really happens and what I have to do to resolve the issue.

If a release needs to be in "SUCCESSFUL" state for helm to accept new upgrades (as it currently seems to be), then the second helm upgrade should explicitly say that: something like: 
> The release is in a "FAILED" state, it cannot be upgraded, please rollback to a "SUCCESSFULL" state first. If not possile, then delete and re-create the release.

It could maybe be done automatically.

Currently as a helm user I have 2 possible behaviors:
- I spotted the error on the first upgrade, fixed it (in values or in the chart definition), and want to apply the fix: I do a second upgrade: it doesn't work, with an unrelated error.
- I did not spot the error on the first upgrade, and try again the upgrade without any change: I get a new error, still unrelated.

It's always confusing.


That being said, without helm and with raw k8s definitions to fix an error I would  just apply everything again, on top of the failed state: couldn't helm do the same? What issue could happen?
